Compilation error on Linux

- [x] I have read [CONTRIBUTING](https://github.com/Quick/Nimble/blob/master/CONTRIBUTING.md) and have done my best to follow them.

### What did you do?

I upgraded my OSS project to use Swift 4 and tried to build the thing on a Travis Linux (Ubuntu 14.04.5) machine. The package manifest for my project:

```swift
let package = Package(
    name: "MockSix",
    dependencies: [
        .package(url: "https://github.com/Quick/Quick.git", from: "1.2.0"),
        .package(url: "https://github.com/Quick/Nimble.git", from: "7.0.2"),
    ],
    targets: [
        .target(name: "MockSix", dependencies: [], path: "Sources"),
        .testTarget(
            name: "MockSixTests",
            dependencies: ["MockSix", "Quick", "Nimble"]
        )
    ],
    swiftLanguageVersions: [4]
)
```

### What did you expect to happen?

`swift build && swift test` to execute properly.

### What actually happened instead?

There was a compilation error while the SPM was building the dependencies:

```
/home/travis/build/lvsti/MockSix/.build/checkouts/Nimble.git--3756397004363768916/Sources/Nimble/Matchers/PostNotification.swift:18:27: error: ambiguous use of 'addObserver'
        self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
                          ^

Foundation.NotificationCenter:9:15: note: found this candidate
    open func addObserver(forName name: Foundation.NSNotification.Name?, object obj: Any?, queue: Foundation.OperationQueue?, usingBlock block: @escaping (Foundation.Notification) -> Swift.Void) -> NSObjectProtocol
              ^

Foundation.NotificationCenter:10:15: note: found this candidate
    open func addObserver(forName name: Foundation.NSNotification.Name?, object obj: Any?, queue: Foundation.OperationQueue?, using block: @escaping (Foundation.Notification) -> Swift.Void) -> NSObjectProtocol
              ^
```

### Environment

List the software versions you're using:

 - Quick: *1.2.0*
 - Nimble: *7.0.2*
 - Xcode Version: N/A
 - Swift Version: *4.0.2* (toolchain: https://swift.org/builds/swift-4.0.2-release/ubuntu1404/swift-4.0.2-RELEASE/swift-4.0.2-RELEASE-ubuntu14.04.tar.gz)
 - Swift Package Manager *Swift 4.0.0-dev*

--- a/Sources/Nimble/Matchers/PostNotification.swift
+++ b/Sources/Nimble/Matchers/PostNotification.swift
@@ -1,5 +1,32 @@
 import Foundation
 
+// A workaround to SR-6419.
+extension NotificationCenter {
+#if !(os(macOS) || os(iOS) || os(tvOS) || os(watchOS))
+    #if swift(>=4.0)
+        #if swift(>=4.0.2)
+        #else
+            // swiftlint:disable:next line_length
+            func addObserver(forName name: Notification.Name?, object obj: Any?, queue: OperationQueue?, using block: @escaping (Notification) -> Void) -> NSObjectProtocol {
+                return addObserver(forName: name, object: obj, queue: queue, usingBlock: block)
+            }
+        #endif
+    #elseif swift(>=3.2)
+        #if swift(>=3.2.2)
+        #else
+            // swiftlint:disable:next line_length
+            func addObserver(forName name: Notification.Name?, object obj: Any?, queue: OperationQueue?, using block: @escaping (Notification) -> Void) -> NSObjectProtocol {
+                return addObserver(forName: name, object: obj, queue: queue, usingBlock: block)
+            }
+        #endif
+    #else
+        func addObserver(forName name: Notification.Name?, object obj: Any?, queue: OperationQueue?, using block: @escaping (Notification) -> Void) -> NSObjectProtocol {
+            return addObserver(forName: name, object: obj, queue: queue, usingBlock: block)
+        }
+    #endif
+#endif
+}
+
 internal class NotificationCollector {
     private(set) var observedNotifications: [Notification]
     private let notificationCenter: NotificationCenter
@@ -15,46 +42,11 @@ internal class NotificationCollector {
     }
 
     func startObserving() {
-        // A workaround to SR-6419.
-        #if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
-            self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
-                // linux-swift gets confused by .append(n)
-                self?.observedNotifications.append(n)
-            }
-        #elseif swift(>=4.0)
-            #if swift(>=4.0.2)
-                // swiftlint:disable:next line_length
-                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil, using: { [weak self] n in
-                    // linux-swift gets confused by .append(n)
-                    self?.observedNotifications.append(n)
-                })
-            #else
-                // swiftlint:disable:next line_length
-                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
-                    // linux-swift gets confused by .append(n)
-                    self?.observedNotifications.append(n)
-                }
-            #endif
-        #elseif swift(>=3.2)
-            #if swift(>=3.2.2)
-                // swiftlint:disable:next line_length
-                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil, using: { [weak self] n in
-                    // linux-swift gets confused by .append(n)
-                    self?.observedNotifications.append(n)
-                })
-            #else
-                // swiftlint:disable:next line_length
-                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
-                    // linux-swift gets confused by .append(n)
-                    self?.observedNotifications.append(n)
-                }
-            #endif
-        #else
-            self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
-                // linux-swift gets confused by .append(n)
-                self?.observedNotifications.append(n)
-            }
-        #endif
+        // swiftlint:disable:next line_length
+        self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil, using: { [weak self] n in
+            // linux-swift gets confused by .append(n)
+            self?.observedNotifications.append(n)
+        })
     }
 
     deinit {

--- a/Sources/Nimble/Matchers/PostNotification.swift
+++ b/Sources/Nimble/Matchers/PostNotification.swift
@@ -15,10 +15,46 @@ internal class NotificationCollector {
     }
 
     func startObserving() {
-        self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
-            // linux-swift gets confused by .append(n)
-            self?.observedNotifications.append(n)
-        }
+        // A workaround to SR-6419.
+        #if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+            self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
+                // linux-swift gets confused by .append(n)
+                self?.observedNotifications.append(n)
+            }
+        #elseif swift(>=4.0)
+            #if swift(>=4.0.2)
+                // swiftlint:disable:next line_length
+                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil, using: { [weak self] n in
+                    // linux-swift gets confused by .append(n)
+                    self?.observedNotifications.append(n)
+                })
+            #else
+                // swiftlint:disable:next line_length
+                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
+                    // linux-swift gets confused by .append(n)
+                    self?.observedNotifications.append(n)
+                }
+            #endif
+        #elseif swift(>=3.2)
+            #if swift(>=3.2.2)
+                // swiftlint:disable:next line_length
+                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil, using: { [weak self] n in
+                    // linux-swift gets confused by .append(n)
+                    self?.observedNotifications.append(n)
+                })
+            #else
+                // swiftlint:disable:next line_length
+                self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
+                    // linux-swift gets confused by .append(n)
+                    self?.observedNotifications.append(n)
+                }
+            #endif
+        #else
+            self.token = self.notificationCenter.addObserver(forName: nil, object: nil, queue: nil) { [weak self] n in
+                // linux-swift gets confused by .append(n)
+                self?.observedNotifications.append(n)
+            }
+        #endif
     }
 
     deinit {

